modules/bucket-events/main.tf

```diff
@@ -0,0 +1,214 @@
+resource "google_pubsub_topic_iam_binding" "allow-pubsub-to-send-to-dead-letter" {
+  for_each = var.regions
+  topic    = google_pubsub_topic.dead-letter.name
+
+  role    = "roles/pubsub.publisher"
+  members = ["serviceAccount:${google_project_service_identity.pubsub.email}"]
+}
+
+// Configure the subscription to deliver the events matching our filter to this service
+// using the above identity to authorize the delivery..
+resource "google_pubsub_subscription" "this" {
+  for_each   = var.regions
+  depends_on = [module.this]
+
+  name  = "${var.name}-${random_string.suffix.result}"
+  topic = google_pubsub_topic.internal[each.key].id
+
+  // TODO: Tune this and/or make it configurable?
+  ack_deadline_seconds = 300
+
+  push_config {
+    push_endpoint = module.authorize-delivery[each.key].uri
+
+    // Authenticate requests to this service using tokens minted
+    // from the given service account.
+    oidc_token {
+      service_account_email = google_service_account.this.email
+    }
+
+    // Make the body of the push notification the raw Pub/Sub message.
+    // Include the Pub/Sub message attributes as HTTP headers.
+    // This aligns the shape of the notification with the "binary"
+    // Cloud Event delivery form.
+    // See: https://cloud.google.com/pubsub/docs/payload-unwrapping
+    no_wrapper {
+      write_metadata = true
+    }
+  }
+
+  expiration_policy {
+    ttl = "" // This does not expire.
+  }
+
+  dead_letter_policy {
+    dead_letter_topic     = google_pubsub_topic.dead-letter.id
+    max_delivery_attempts = var.max_delivery_attempts
+  }
+}
+
+// Grant the pubsub service account the ability to Acknowledge messages on this "this" subscription.
+resource "google_pubsub_subscription_iam_binding" "allow-pubsub-to-ack" {
+  for_each     = var.regions
+  subscription = google_pubsub_subscription.this[each.key].name
+
+  role    = "roles/pubsub.subscriber"
+  members = ["serviceAccount:${google_project_service_identity.pubsub.email}"]
+}
+
+// Lookup the GCS service account.
+data "google_storage_project_service_account" "gcs_account" {}
+
+// Allow the GCS service account to publish to the internal topic.
+resource "google_pubsub_topic_iam_binding" "binding" {
+  for_each = var.regions
+  topic    = google_pubsub_topic.internal[each.key].id
+  role     = "roles/pubsub.publisher"
+  members  = ["serviceAccount:${data.google_storage_project_service_account.gcs_account.email_address}"]
+}
+
+// Creage the topic to receive the GCS events.
+resource "google_pubsub_topic" "internal" {
+  for_each = var.regions
+  name     = "${var.name}-internal-${each.key}"
+
+  // TODO: Tune this and/or make it configurable?
+  message_retention_duration = "600s"
+
+  message_storage_policy {
+    allowed_persistence_regions = [each.key]
+  }
+}
+
+// Create a notification to the topic.
+resource "google_storage_notification" "notification" {
+  for_each       = var.regions
```

Comment:
You don't want this.  It will mean you get N of the event instead of 1.  I think you want the single region colocated with the bucket.

Reply:
Okay that makes sense. So you'd expect this module to take a us-central1 bucket and create a us-central1 topic/subscription/service that publishes to an ingress also only in that region?

I had started more along those lines, but thought we'd want to unshackle ourselves from regions as soon as possible.

But I can go back that way too. For multi-region buckets (`US`, `EU`) do we want multi-regional subscriber setups?

Reply:
I think that's what we want.  It makes data sovereignty considerations a lot simpler and practically speaking with geographic load balancing it'd hit the colocated service anyways, or (like this) we will do Nx the amount of work because EVERY region will get a subscription event instead of just the one the bucket lives in.

re: multi-regional buckets

With egress cost becoming a thing for muilti-regional access within that multi-region, I've been gravitating towards regional storage for basically everything.